### ISSUE TITLE
Pyright ignoring explicit types when creating an optional type which is initiated as `None`

### ISSUE THREAD

**Issue by vidhanio** [microsoft/pyright#4023]
**Describe the bug**
Pyright seems to ignore explicit types when creating an optional type which starts off as `None`, but then gets assigned to in a `for` loop.

**To Reproduce**
```py
from typing import NamedTuple


class Calculation(NamedTuple):
    decimal: float
    numerator: int
    denominator: int


def approximation(actual: float, denominator: int) -> tuple[int, int]:
    closest: Calculation | None = None

    for num in range(1, denominator + 1):
        for denom in range(1, denominator + 1):
            decimal = num / denom
            if decimal < actual:
                if closest is None:
                    closest = Calculation(decimal, num, denom)
                elif closest.decimal < decimal:
                        closest = Calculation(decimal, num, denom)

    return (0, 0) if closest is None else (closest.numerator, closest.denominator) # tuple[Literal[0], Literal[0]] | tuple[Unknown, Unknown]
```

**Expected behavior**
The type of the return should be `tuple[int, int]`

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
<img width="838" alt="image" src="https://user-images.githubusercontent.com/41439633/194685208-e0ba51f4-d88c-472a-b7d7-21bcbafa7de3.png">

**VS Code extension or command-line**
Pyright is being run through Pylance in VS Code.

**Comment by erictraut** [comment]
Thanks for the bug report.

Here's a simplified example:

```python
from typing import NamedTuple

class Calculation(NamedTuple):
    x: float

def func1():
    a: Calculation | None = None

    for _ in range(1):
        for _ in range(1):
            a = Calculation(1)
        reveal_type(a, expected_type=Calculation | None)
    reveal_type(a, expected_type=Calculation | None)
```

I'll investigate further.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by vidhanio** [comment]
Thank you so much for the update and extremely fast fix! :)

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.275, which I just published. It will also be included in a future release of pylance.
